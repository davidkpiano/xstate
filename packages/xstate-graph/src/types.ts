import {
  State,
  EventObject,
  StateValue,
  StateNode,
  TransitionDefinition,
  MachineContext
} from 'xstate';

export interface TransitionMap {
  state: StateValue | undefined;
}

export type JSONSerializable<T extends object, U> = T & {
  toJSON: () => U;
};

export type DirectedGraphLabel = JSONSerializable<
  {
    text: string;
  },
  {
    text: string;
  }
>;

export type DirectedGraphEdge = JSONSerializable<
  {
    id: string;
    source: StateNode;
    target: StateNode;
    label: DirectedGraphLabel;
    transition: TransitionDefinition<any, any>;
  },
  {
    source: string;
    target: string;
    label: ReturnType<DirectedGraphLabel['toJSON']>;
  }
>;

// Based on https://www.eclipse.org/elk/documentation/tooldevelopers/graphdatastructure/jsonformat.html
export type DirectedGraphNode = JSONSerializable<
  {
    id: string;
    stateNode: StateNode;
    children: DirectedGraphNode[];
    /**
     * The edges representing all transitions from this `stateNode`.
     */
    edges: DirectedGraphEdge[];
  },
  {
    id: string;
    children: DirectedGraphNode[];
  }
>;

export interface AdjacencyMap<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  [stateId: string]: Record<
    string,
    {
      state: State<TContext, TEvent>;
      event: TEvent;
    }
  >;
}

export interface StatePaths<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  /**
   * The target state.
   */
  state: State<TContext, TEvent>;
  /**
   * The paths that reach the target state.
   */
  paths: Array<StatePath<TContext, TEvent>>;
}

export interface StatePath<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  /**
   * The ending state of the path.
   */
  state: State<TContext, TEvent>;
  /**
   * The ordered array of state-event pairs (segments) which reach the ending `state`.
   */
  segments: Segments<TContext, TEvent>;
  /**
   * The combined weight of all segments in the path.
   */
  weight: number;
}

export interface StatePathsMap<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  [key: string]: StatePaths<TContext, TEvent>;
}
export interface Segment<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  /**
   * The current state before taking the event.
   */
  state: State<TContext, TEvent>;
  /**
   * The event to be taken from the specified state.
   */
  event: TEvent;
}

export type Segments<
  TContext extends MachineContext,
  TEvent extends EventObject
> = Array<Segment<TContext, TEvent>>;

export type ExtractEvent<
  TEvent extends EventObject,
  TType extends TEvent['type']
> = TEvent extends { type: TType } ? TEvent : never;

export interface ValueAdjMapOptions<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  events?: {
    [K in TEvent['type']]?:
      | Array<ExtractEvent<TEvent, K>>
      | ((state: State<TContext, TEvent>) => Array<ExtractEvent<TEvent, K>>);
  };
  filter?: (state: State<TContext, any>) => boolean;
  stateSerializer?: (state: State<TContext, any>) => string;
  eventSerializer?: (event: TEvent) => string;
}
