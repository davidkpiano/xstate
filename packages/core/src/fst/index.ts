// https://en.wikipedia.org/wiki/Finite-state_transducer

import type { State, StateMachine } from '..';

export { toActor as fromActor } from './actor';
export { fromMachine } from './machine';
export { fromReducer } from './reducer';

export type FSTTransition<TState, TEvent, TOutput> = (
  state: TState,
  input: TEvent
) => [TState] | [TState, TOutput];

export interface FST<TState, TEvent, TOutput = any> {
  transition: FSTTransition<TState, TEvent, TOutput>;
  initialState: TState;
  nextEvents?: (state: TState) => TEvent[];
}

export type StateTransducerFrom<T> = T extends StateMachine<
  infer TContext,
  any,
  infer TEvent
>
  ? FST<State<TContext, TEvent>, TEvent, any[]>
  : never;
