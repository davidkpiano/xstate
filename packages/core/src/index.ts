import { matchesState } from './utils';
import { mapState } from './mapState';
import { StateNode } from './StateNode';
import { State } from './State';
import { createMachine } from './Machine';
import {
  raise,
  send,
  sendParent,
  sendUpdate,
  log,
  cancel,
  stop,
  assign,
  after,
  done,
  respond,
  doneInvoke,
  forwardTo,
  escalate,
  choose,
  pure
} from './actions';
import { interpret, Interpreter, InterpreterStatus } from './interpreter';
import { matchState } from './match';
export { StateMachine as MachineNode } from './StateMachine';
export { SimulatedClock } from './SimulatedClock';
export {
  spawn,
  spawnFrom,
  spawnMachine,
  spawnPromise,
  spawnObservable,
  spawnCallback
} from './actor';
export { createSchema } from './schema';

const actions = {
  raise,
  send,
  sendParent,
  sendUpdate,
  log,
  cancel,
  stop,
  assign,
  after,
  done,
  respond,
  forwardTo,
  escalate,
  choose,
  pure
};

export {
  StateNode,
  State,
  matchesState,
  mapState,
  actions,
  assign,
  send,
  sendParent,
  sendUpdate,
  forwardTo,
  interpret,
  Interpreter,
  InterpreterStatus,
  matchState,
  doneInvoke,
  createMachine
};

export * from './types';

// TODO: decide from where those should be exported
export { pathToStateValue, flatten, keys } from './utils';
export { getStateNodes } from './stateUtils';
