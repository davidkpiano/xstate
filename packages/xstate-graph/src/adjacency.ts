import {
  DefaultContext,
  Event,
  EventObject,
  StateMachine,
  AnyEventObject,
  State
} from 'xstate';
import { flatten } from 'xstate/lib/utils';
import { FST, fromMachine } from 'xstate/lib/fst';
import { AdjacencyMap, ValueAdjMapOptions, AdjacencyMapFST } from './types';
import { defaultValueAdjMapOptions, toEventObject } from './graph';

export function nextEventsGetter<TContext, TEvent extends EventObject>(
  machine: StateMachine<TContext, any, TEvent>,
  events: ValueAdjMapOptions<TContext, TEvent>['events'] = {}
): (state: State<TContext, TEvent>) => TEvent[] {
  const allEvents = {} as Record<TEvent['type'], Array<Event<TEvent>>>;
  for (const event of machine.events) {
    allEvents[event] = [event];
  }
  Object.assign(allEvents, events);

  return (state) => {
    const { nextEvents } = state;
    const potentialEvents = flatten<TEvent>(
      nextEvents.map((nextEvent) => {
        const getNextEvents = events[nextEvent];

        if (!getNextEvents) {
          return [{ type: nextEvent }];
        }

        if (typeof getNextEvents === 'function') {
          return getNextEvents(state);
        }

        return getNextEvents;
      })
    ).map((event) => toEventObject(event));

    return potentialEvents;
  };
}

function getValueAdjMapOptions<TContext, TEvent extends EventObject>(
  options?: ValueAdjMapOptions<TContext, TEvent>
): Required<ValueAdjMapOptions<TContext, TEvent>> {
  return {
    ...(defaultValueAdjMapOptions as Required<
      ValueAdjMapOptions<TContext, TEvent>
    >),
    ...options
  };
}

export function getAdjacencyMap<
  TContext = DefaultContext,
  TEvent extends EventObject = AnyEventObject
>(
  machine: StateMachine<TContext, any, TEvent>,
  options?: ValueAdjMapOptions<TContext, TEvent>
): AdjacencyMap<TContext, TEvent> {
  const optionsWithDefaults = getValueAdjMapOptions(options);

  const fst = fromMachine(
    machine,
    nextEventsGetter(machine, optionsWithDefaults.events as any)
  );

  const adjacency = getAdjacencyMapFST(fst, options);

  return adjacency as AdjacencyMap<TContext, TEvent>;
}

export function getAdjacencyMapFST<TState, TInput>(
  fst: FST<TState, TInput>,
  options?: Partial<ValueAdjMapOptions<any, any>>
): AdjacencyMapFST<TState, TInput> {
  const optionsWithDefaults = getValueAdjMapOptions(options);
  const { filter, stateSerializer, eventSerializer } = optionsWithDefaults;

  const adjacency: AdjacencyMapFST<TState, TInput> = {};

  function findAdjacencies(state: TState) {
    const potentialEvents = fst.nextEvents?.(state) ?? [];
    const stateHash = stateSerializer(state as any);

    if (adjacency[stateHash]) {
      return;
    }

    adjacency[stateHash] = {};

    for (const event of potentialEvents) {
      let nextState: TState;
      try {
        nextState = fst.transition(state, event)[0];
      } catch (e) {
        throw new Error(
          `Unable to transition from state ${stateSerializer(
            state as any
          )} on event ${eventSerializer(event)}: ${e.message}`
        );
      }

      if (
        (!filter || filter(nextState as any)) &&
        stateHash !== stateSerializer(nextState as any)
      ) {
        adjacency[stateHash][eventSerializer(event)] = {
          state: nextState,
          event
        };

        findAdjacencies(nextState);
      }
    }
  }

  findAdjacencies(fst.initialState);

  return adjacency;
}
