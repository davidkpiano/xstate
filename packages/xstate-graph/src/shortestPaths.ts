import { DefaultContext, EventObject, StateMachine } from 'xstate';
import { isMachine, keys } from 'xstate/lib/utils';
import {
  StatePathsMap,
  StatePathsMapFST,
  ValueAdjMapOptions,
  ValueAdjMapOptionsFST
} from './types';
import { getAdjacencyMapFST, nextEventsGetter } from './adjacency';
import {
  EMPTY_MAP,
  serializeState,
  serializeEvent,
  deserializeEventString
} from './graph';
import { FST, fromMachine } from 'xstate/lib/fst';

export function getShortestPaths<
  TContext = DefaultContext,
  TEvent extends EventObject = EventObject
>(
  machine: StateMachine<TContext, any, TEvent>,
  options?: Partial<ValueAdjMapOptions<TContext, TEvent>>
): StatePathsMap<TContext, TEvent> {
  if (isMachine(machine) && !machine.states) {
    return EMPTY_MAP;
  }
  const optionsWithDefaults = {
    events: {},
    stateSerializer: serializeState,
    eventSerializer: serializeEvent,
    ...options
  } as ValueAdjMapOptions<TContext, TEvent>;

  const fst = fromMachine(
    machine,
    nextEventsGetter(machine, optionsWithDefaults.events as any)
  );

  const statePathMap = getShortestPathsFST(fst, optionsWithDefaults as any);

  return statePathMap as StatePathsMap<TContext, TEvent>;
}

export function getShortestPathsFST<TState, TInput>(
  fst: FST<TState, TInput>,
  options?: Partial<ValueAdjMapOptionsFST<TState, TInput>>
): StatePathsMapFST<TState, TInput> {
  const optionsWithDefaults = {
    events: {},
    stateSerializer: serializeState,
    eventSerializer: serializeEvent,
    ...options
  } as any;

  const adjacency = getAdjacencyMapFST<TState, TInput>(
    fst,
    optionsWithDefaults
  );

  // weight, state, event
  const weightMap = new Map<
    string,
    [number, string | undefined, string | undefined]
  >();
  const stateMap = new Map<string, TState>();
  const initialVertex = optionsWithDefaults.stateSerializer(
    fst.initialState as any
  );
  stateMap.set(initialVertex, fst.initialState);

  weightMap.set(initialVertex, [0, undefined, undefined]);
  const unvisited = new Set<string>();
  const visited = new Set<string>();

  unvisited.add(initialVertex);
  while (unvisited.size > 0) {
    for (const vertex of unvisited) {
      const [weight] = weightMap.get(vertex)!;
      for (const event of keys(adjacency[vertex])) {
        const nextSegment = adjacency[vertex][event];
        const nextVertex = optionsWithDefaults.stateSerializer(
          nextSegment.state as any
        );
        stateMap.set(nextVertex, nextSegment.state);
        if (!weightMap.has(nextVertex)) {
          weightMap.set(nextVertex, [weight + 1, vertex, event]);
        } else {
          const [nextWeight] = weightMap.get(nextVertex)!;
          if (nextWeight > weight + 1) {
            weightMap.set(nextVertex, [weight + 1, vertex, event]);
          }
        }
        if (!visited.has(nextVertex)) {
          unvisited.add(nextVertex);
        }
      }
      visited.add(vertex);
      unvisited.delete(vertex);
    }
  }

  const statePathMap: StatePathsMapFST<TState, TInput> = {};

  weightMap.forEach(([weight, fromState, fromEvent], stateSerial) => {
    const state = stateMap.get(stateSerial)!;
    statePathMap[stateSerial] = {
      state,
      paths: !fromState
        ? [
            {
              state,
              segments: [],
              weight
            }
          ]
        : [
            {
              state,
              segments: statePathMap[fromState].paths[0].segments.concat({
                state: stateMap.get(fromState)!,
                event: (deserializeEventString(fromEvent!) as unknown) as TInput
              }),
              weight
            }
          ]
    };
  });

  return statePathMap;
}
