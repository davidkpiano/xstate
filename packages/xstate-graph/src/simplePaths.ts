import { DefaultContext, EventObject, StateMachine } from 'xstate';
import { isMachine, keys } from 'xstate/lib/utils';
import {
  StatePathsMap,
  ValueAdjMapOptions,
  StatePathsMapFST,
  SegmentsFST,
  ValueAdjMapOptionsFST
} from './types';
import { getAdjacencyMapFST, nextEventsGetter } from './adjacency';
import {
  defaultValueAdjMapOptions,
  EMPTY_MAP,
  deserializeEventString
} from './graph';
import { FST, fromMachine } from 'xstate/lib/fst';

export function getSimplePaths<
  TContext = DefaultContext,
  TEvent extends EventObject = EventObject
>(
  machine: StateMachine<TContext, any, TEvent>,
  options?: Partial<ValueAdjMapOptions<TContext, TEvent>>
): StatePathsMap<TContext, TEvent> {
  const optionsWithDefaults = {
    ...defaultValueAdjMapOptions,
    ...options
  };

  if (isMachine(machine) && !machine.states) {
    return EMPTY_MAP;
  }

  const fst = fromMachine(
    machine,
    nextEventsGetter(machine, optionsWithDefaults.events as any)
  );

  const paths = getSimplePathsFST(fst, optionsWithDefaults as any);

  return paths as StatePathsMap<TContext, TEvent>;
}

export function getSimplePathsFST<TState, Tinput>(
  fst: FST<TState, Tinput>,
  options?: Partial<ValueAdjMapOptionsFST<TState, Tinput>>
): StatePathsMapFST<TState, Tinput> {
  const optionsWithDefaults = {
    ...defaultValueAdjMapOptions,
    ...options
  };

  const { stateSerializer } = optionsWithDefaults;
  // @ts-ignore - excessively deep
  const adjacency = getAdjacencyMapFST(fst, optionsWithDefaults);
  const stateMap = new Map<string, TState>();
  const visited = new Set();
  const path: SegmentsFST<TState, Tinput> = [];
  const paths: StatePathsMapFST<TState, Tinput> = {};

  function util(fromState: TState, toStateSerial: string) {
    const fromStateSerial = stateSerializer(fromState as any);
    visited.add(fromStateSerial);

    if (fromStateSerial === toStateSerial) {
      if (!paths[toStateSerial]) {
        paths[toStateSerial] = {
          state: stateMap.get(toStateSerial)!,
          paths: []
        };
      }
      paths[toStateSerial].paths.push({
        state: fromState,
        weight: path.length,
        segments: [...path]
      });
    } else {
      for (const subEvent of keys(adjacency[fromStateSerial])) {
        const nextSegment = adjacency[fromStateSerial][subEvent];

        if (!nextSegment) {
          continue;
        }

        const nextStateSerial = stateSerializer(nextSegment.state as any);
        stateMap.set(nextStateSerial, nextSegment.state);

        if (!visited.has(nextStateSerial)) {
          path.push({
            state: stateMap.get(fromStateSerial)!,
            event: deserializeEventString(subEvent) as any
          });
          util(nextSegment.state, toStateSerial);
        }
      }
    }

    path.pop();
    visited.delete(fromStateSerial);
  }

  const initialStateSerial = stateSerializer(fst.initialState as any);
  stateMap.set(initialStateSerial, fst.initialState);

  for (const nextStateSerial of keys(adjacency)) {
    util(fst.initialState, nextStateSerial);
  }

  return paths;
}
