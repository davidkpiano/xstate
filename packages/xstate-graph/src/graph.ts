import {
  StateNode,
  State,
  DefaultContext,
  Event,
  EventObject,
  StateMachine
} from 'xstate';
import { flatten, keys } from 'xstate/lib/utils';
import {
  StatePaths,
  ValueAdjMapOptions,
  DirectedGraphEdge,
  DirectedGraphNode
} from './types';
import { getSimplePaths } from './simplePaths';

export function toEventObject<TEvent extends EventObject>(
  event: Event<TEvent>
): TEvent {
  if (typeof event === 'string' || typeof event === 'number') {
    return ({ type: event } as unknown) as TEvent;
  }

  return event;
}

export const EMPTY_MAP = {};

/**
 * Returns all state nodes of the given `node`.
 * @param stateNode State node to recursively get child state nodes from
 */
export function getStateNodes(
  stateNode: StateNode | StateMachine<any, any, any>
): StateNode[] {
  const { states } = stateNode;
  const nodes = keys(states).reduce((accNodes: StateNode[], stateKey) => {
    const childStateNode = states[stateKey];
    const childStateNodes = getStateNodes(childStateNode);

    accNodes.push(childStateNode, ...childStateNodes);
    return accNodes;
  }, []);

  return nodes;
}

export function getChildren(stateNode: StateNode): StateNode[] {
  if (!stateNode.states) {
    return [];
  }

  const children = Object.keys(stateNode.states).map((key) => {
    return stateNode.states[key];
  });

  return children;
}

export function serializeState<TContext>(state: State<TContext, any>): string {
  const { value, context } = state;
  return context === undefined
    ? JSON.stringify(value)
    : JSON.stringify(value) + ' | ' + JSON.stringify(context);
}

export function serializeEvent<TEvent extends EventObject>(
  event: TEvent
): string {
  return JSON.stringify(event);
}

export function deserializeEventString<TEvent extends EventObject>(
  eventString: string
): TEvent {
  return JSON.parse(eventString) as TEvent;
}

export const defaultValueAdjMapOptions: Required<ValueAdjMapOptions<
  any,
  any
>> = {
  events: {},
  filter: () => true,
  stateSerializer: serializeState,
  eventSerializer: serializeEvent
};

export function getSimplePathsAsArray<
  TContext = DefaultContext,
  TEvent extends EventObject = EventObject
>(
  machine: StateMachine<TContext, any, TEvent>,
  options?: ValueAdjMapOptions<TContext, TEvent>
): Array<StatePaths<TContext, TEvent>> {
  const result = getSimplePaths(machine, options);
  return keys(result).map((key) => result[key]);
}

export function toDirectedGraph(stateNode: StateNode): DirectedGraphNode {
  const edges: DirectedGraphEdge[] = flatten(
    stateNode.transitions.map((t, transitionIndex) => {
      const targets = t.target ? t.target : [stateNode];

      return targets.map((target, targetIndex) => {
        const edge: DirectedGraphEdge = {
          id: `${stateNode.id}:${transitionIndex}:${targetIndex}`,
          source: stateNode,
          target,
          transition: t,
          label: {
            text: t.eventType,
            toJSON: () => ({ text: t.eventType })
          },
          toJSON: () => {
            const { label } = edge;

            return { source: stateNode.id, target: target.id, label };
          }
        };

        return edge;
      });
    })
  );

  const graph = {
    id: stateNode.id,
    stateNode,
    children: getChildren(stateNode).map((sn) => toDirectedGraph(sn)),
    edges,
    toJSON: () => {
      const { id, children, edges: graphEdges } = graph;
      return { id, children, edges: graphEdges };
    }
  };

  return graph;
}
