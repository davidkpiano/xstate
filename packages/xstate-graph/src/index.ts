import {
  getStateNodes,
  serializeEvent,
  serializeState,
  toDirectedGraph
} from './graph';
import { getSimplePaths } from './simplePaths';
import { getShortestPaths } from './shortestPaths';

export {
  getStateNodes,
  getSimplePaths,
  getShortestPaths,
  serializeEvent,
  serializeState,
  toDirectedGraph
};

export * from './types';
