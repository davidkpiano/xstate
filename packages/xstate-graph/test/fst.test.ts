import { FST } from 'xstate/lib/fst';
import { getShortestPathsFST } from '../src/shortestPaths';

describe('fst', () => {
  it('should', () => {
    class Jugs {
      public three = 0;
      public five = 0;

      public fillThree() {
        this.three = 3;
      }
      public fillFive() {
        this.five = 5;
      }
      public emptyThree() {
        this.three = 0;
      }
      public emptyFive() {
        this.five = 0;
      }
      public transferThree() {
        const poured = Math.min(5 - this.five, this.three);

        this.three = this.three - poured;
        this.five = this.five + poured;
      }
      public transferFive() {
        const poured = Math.min(3 - this.three, this.five);

        this.three = this.three + poured;
        this.five = this.five - poured;
      }
    }

    const dieHardFst: FST<
      [number, number],
      | { type: 'FILL_3' }
      | { type: 'FILL_5' }
      | { type: 'EMPTY_3' }
      | { type: 'EMPTY_5' }
      | { type: 'TRANSFER_3' }
      | { type: 'TRANSFER_5' }
    > = {
      transition: (gallons, event) => {
        const jugs = new Jugs();
        [jugs.three, jugs.five] = gallons;

        switch (event.type) {
          case 'EMPTY_3':
            jugs.emptyThree();
            break;
          case 'EMPTY_5':
            jugs.emptyFive();
            break;
          case 'FILL_3':
            jugs.fillThree();
            break;
          case 'FILL_5':
            jugs.fillFive();
            break;
          case 'TRANSFER_3':
            jugs.transferThree();
            break;
          case 'TRANSFER_5':
            jugs.transferFive();
            break;
          default:
            break;
        }

        return [[jugs.three, jugs.five]];
      },
      initialState: [0, 0],
      nextEvents: () => [
        { type: 'FILL_3' },
        { type: 'FILL_5' },
        { type: 'EMPTY_3' },
        { type: 'EMPTY_5' },
        { type: 'TRANSFER_3' },
        { type: 'TRANSFER_5' }
      ]
    };

    const shortestPaths = getShortestPathsFST(dieHardFst, {
      stateSerializer: ([three, five]) => JSON.stringify([three, five])
    });

    const pathsToFour = Object.values(shortestPaths).filter(({ state }) => {
      return state[1] === 4;
    });

    expect(pathsToFour).toMatchSnapshot();
  });
});
