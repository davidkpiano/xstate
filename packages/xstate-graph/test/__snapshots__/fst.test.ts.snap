// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fst should 1`] = `
Array [
  Object {
    "paths": Array [
      Object {
        "segments": Array [
          Object {
            "event": Object {
              "type": "FILL_5",
            },
            "state": Array [
              0,
              0,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              0,
              5,
            ],
          },
          Object {
            "event": Object {
              "type": "EMPTY_3",
            },
            "state": Array [
              3,
              2,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              0,
              2,
            ],
          },
          Object {
            "event": Object {
              "type": "FILL_5",
            },
            "state": Array [
              2,
              0,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              2,
              5,
            ],
          },
        ],
        "state": Array [
          3,
          4,
        ],
        "weight": 6,
      },
    ],
    "state": Array [
      3,
      4,
    ],
  },
  Object {
    "paths": Array [
      Object {
        "segments": Array [
          Object {
            "event": Object {
              "type": "FILL_5",
            },
            "state": Array [
              0,
              0,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              0,
              5,
            ],
          },
          Object {
            "event": Object {
              "type": "EMPTY_3",
            },
            "state": Array [
              3,
              2,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              0,
              2,
            ],
          },
          Object {
            "event": Object {
              "type": "FILL_5",
            },
            "state": Array [
              2,
              0,
            ],
          },
          Object {
            "event": Object {
              "type": "TRANSFER_5",
            },
            "state": Array [
              2,
              5,
            ],
          },
          Object {
            "event": Object {
              "type": "EMPTY_3",
            },
            "state": Array [
              3,
              4,
            ],
          },
        ],
        "state": Array [
          0,
          4,
        ],
        "weight": 7,
      },
    ],
    "state": Array [
      0,
      4,
    ],
  },
]
`;
